keps/sig-cluster-lifecycle/0015-kubeadm-join-master.md

```diff
@@ -0,0 +1,431 @@
+designed with the objective to introduce an upstream simple and reliable solution for
+achieving the same goal.
+
+Such solution will provide a consistent and repeatable base for implementing additional
+capabilities like e.g. kubeadm upgrade for HA clusters.
+
+### Goals
+
+- "Divide and conquer”
+
+  This proposal - at least in its initial release - does not address all the possible
+  user stories for creating an highly available Kubernetes cluster, but instead
+  focuses on:
+
+  - Defining a generic and extensible flow for bootstrapping an HA cluster, the
+    `kubeadm join --master` workflow.
+  - Providing a solution *only* for well defined user stories. see
+    [User Stories](#user-stories) and [Non-goals](#non-goals).
+
+- Enable higher-level tools integration
+
+  We expect higher-level and tooling will leverage on kubeadm for creating HA clusters;
+  accordingly, the `kubeadm join --master` workflow should provide support for
+  the following operational practices used by higher level tools:
+
+  - Parallel node creation
+
+    Higher-level tools could create nodes in parallel (both masters and workers)
+    for reducing the overall cluster startup time.
+    `kubeadm join --master` should support natively this practice without requiring
+    the implementation of any synchronization mechanics by higher-level tools.
+
+  - Replace reconciliation strategies
+
+    Especially in case of cloud deployments, higher-level automation tools could
+    decide for any reason to replace existing nodes with new ones (instead of apply
+    changes in-place to existing nodes).
+    `kubeadm join --master` will support this practice by making easier to replace
+    existing master nodes with new ones.
+
+- Provide support both for dynamic and static bootstrap flow
+
+  At the time a user is running `kubeadm init`, s/he/an operator might not know what
+  the cluster setup will look like eventually. For instance, the user may start with
+  only one master + n nodes, and then add further master nodes with `kubeadm join --master`
+  or add more worker nodes with `kubeadm join` (in any order). This kind of workflow, where the
+  user doesn’t know in advance the final layout of the control plane instances, into this
+  document is referred as “dynamic bootstrap workflow”.
+
+  Nevertheless, kubeadm should support also more “static bootstrap flow”, where a user knows
+  in advance the target layout of the controlplane instances (the number, the name and the IP
+  of master nodes).
+
+### Non-goals
+
+- This proposal doesn't include a solution for etcd cluster management (but nothing in this proposal should
+  prevent to address this in future).
+
+  > At the time of writing, the CoreOS recommended approach for etcd is to run
+  > the etcd cluster outside kubernetes (see discussion in [kubeadm office hours](https://goo.gl/fjyeqo)).
+
+- Nothing in this proposal should prevent users to run master nodes components
```

Comment:
why is this in non-goals? I'd say this is a goal -- to be able to co-locate etcd and the API server _as well_